File: src/predictor.js
// Cycle timer + calibrated odds (weather-style: exact clock, honest dice).
export function anchorFromTimestamp(ts, cycleSeconds = 300) {
  return ts - (ts % cycleSeconds);
}

export function nextReset({ now = Date.now() / 1000, epoch = null, cycleSeconds = 300 } = {}) {
  if (epoch === null || epoch === undefined) {
    const elapsed = now % cycleSeconds;
    return { nextResetIn: cycleSeconds - elapsed, cycleNumber: 0, anchored: false };
  }
  const elapsed = now - epoch;
  return {
    nextResetIn: cycleSeconds - (elapsed % cycleSeconds),
    cycleNumber: Math.floor(elapsed / cycleSeconds) + 1,
    anchored: true,
  };
}

export function countdown(seconds) {
  const m = Math.floor(seconds / 60);
  const s = Math.floor(seconds % 60);
  return `${m}:${String(s).padStart(2, "0")}`;
}

export function oddsText(odds) {
  const lines = [];
  for (const rarity of Object.keys(odds).sort((a, b) => odds[b] - odds[a])) {
    const p = odds[rarity];
    if (p >= 0.001) {
      lines.push(`**${title(rarity)}** ~${+(p * 100)}% per cycle`);
    } else {
      lines.push(`**${title(rarity)}** ~1 in ${Math.round(1 / p).toLocaleString("en-US")} cycles`);
    }
  }
  return lines.join("\n");
}

const title = (s) => s[0].toUpperCase() + s.slice(1);

File: common/predictor.py
"""Cycle timer + calibrated rarity odds.

Two honest kinds of prediction:
1. Timing — the map's eggs reset on a fixed cycle (default 5 min). Once we've
   seen a single real capture we can anchor the grid and count down exactly.
2. Rarity — per-cycle odds per server from the config table. Output is
   weather-forecast style: confident about the clock, honest about the dice.
"""
from __future__ import annotations

import time
from dataclasses import dataclass


@dataclass
class Prediction:
    next_reset_in: int          # seconds until next reset
    cycle_number: int
    anchored: bool

    def countdown(self) -> str:
        m, s = divmod(int(self.next_reset_in), 60)
        return f"{m}:{s:02d}"


def anchor_from_timestamp(ts: float, cycle_seconds: int) -> int:
    """Snap an observed capture time onto the cycle grid; use as epoch."""
    return int(ts - (ts % cycle_seconds))


def next_reset(now: float | None = None, epoch: float | None = None,
               cycle_seconds: int = 300) -> Prediction:
    now = time.time() if now is None else now
    if epoch is None:
        # Not anchored yet: count down on a provisional grid from boot time.
        epoch = 0
        elapsed = now % cycle_seconds
        return Prediction(cycle_seconds - elapsed, 0, anchored=False)
    elapsed = now - epoch
    cycle_number = int(elapsed // cycle_seconds) + 1
    remaining = cycle_seconds - (elapsed % cycle_seconds)
    return Prediction(remaining, cycle_number, anchored=True)


def odds_text(odds: dict[str, float]) -> str:
    """Human-readable per-cycle odds, highest tier first."""
    lines = []
    for rarity in sorted(odds, key=lambda r: odds[r], reverse=True):
        p = odds[rarity]
        if p >= 0.001:
            lines.append(f"**{rarity.title()}** ~{p * 100:g}% per cycle")
        else:
            lines.append(f"**{rarity.title()}** ~1 in {int(round(1 / p)):,} cycles")
    return "\n".join(lines)
